Introduce docs on "Time travel and historical data retention periods" for RETAIN HISTORY feature (#26623)
MaterializeInc/materialize: doc/user/content/transform-data/patterns/time-travel-queries.md

---
title: "Time travel queries"
description: "How to use history retention to enable time travel queries in Materialize"
menu:
  main:
    parent: 'sql-patterns'
    name: "Time travel queries"
---

{{< private-preview />}}

By default, all user-defined sources, tables, materialized views, and indexes
keep track of the most recent version of their underlying data. Some use cases
require accessing past versions of this data at specific points in time
(i.e. time traveling). To allow time travel queries, you can configure these
objects to **retain history**.

When set to a value different than the default (i.e. no history retention), the
history retention period of an object determines the duration for which
historical versions of its underlying data are available for querying.

The common use cases for time traveling in Materialize are:

> * Lossless, **durable subscriptions** to your changing results. See the

Reviewer: Perhaps as a followup: in the past we've talked about organizations that can't handle the continuous nature of SUBSCRIBE. We provide DECLARE + FETCH as a way around this. Using that should be a third bullet point here, with fuller descriptions maybe on the FETCH page.

Author: Agreed! If you don't mind, I'll mark that as a follow up since we're already covering so much in this initial pass.

Author: Added a todo in the docs as a placeholder!

* **For sources, tables and materialized views:** increasing the history
    retention period for these objects increases the amount of historical data
    that is retained in the storage layer. You can expect storage resource
    utilization to increase, which may incur additional costs.

* **For indexes:** increasing the history retention period for an index
    increases the amount of historical data that is retained in memory in the
    cluster maintaining the index. You can expect memory resource utilization
    for that cluster to go up, which might require you to size up your cluster
    and consume additional compute credits.

### Clean-up

The history retention period represents the minimum amount of historical data
guaranteed to be retained by Materialize. History clean-up is processed in the
background, so older history may be accessible for the period of time between
when it falls outside the retention period and when it is cleaned up.

### Best practices

Given the additional memory costs associated with increasing the history
retention period on indexes, if you don't need the performance benefits of an

> index - in particular for the use case of [durable subscriptions](/sql/subscribe#durable-subscriptions) -

Reviewer: ```suggestion
index—in particular for the use case of [durable subscriptions](/sql/subscribe#durable-subscriptions)—
```

em dashes!

- you should consider creating a materialized view for your subscription query
and configure the history retention period on that materialized view
instead.

Similarly, because of the increased storage costs and processing time for the
additional historical data on whichever object has an increased history
retention period, you should consider configuring history retention period on
the index or materialized view directly powering the subscription, rather than
all the way through the dependency chain from the source to the index or
materialized view.

> ## Examples

Reviewer 2: Thoughts from a second pass: since we're spinning this as a pattern, it feels like the how-to for point-in-time queries and durable subscriptions should live here, rather than being buried in the reference documentation. Also, we need to bring `AS OF` back to the [`SELECT`](https://materialize.com/docs/sql/select/) reference page.

Eventually, I think we'll want to extricate the SQL reference bits as a shortcode that can be applied to the low-level reference pages, and focus this page on just guidance on how to implement the time traveling patterns in practice.

Reviewer: Concur.

Author: Just pushed up the change to move the main durable subscription guide to the time travel docs, and put a todo for point in time queries docs.

### Enabling history retention

To set a history retention period for an object, use the `RETAIN HISTORY`
option, which accepts positive [interval](/sql/types/interval/) values
(e.g. `'1hr'`):

```sql
CREATE MATERIALIZED VIEW winning_bids
WITH (RETAIN HISTORY FOR '1hr') AS
SELECT auction_id,
       bid_id,
       item,
       amount
FROM highest_bid_per_auction
WHERE end_time < mz_now();
```

To adjust the history retention period for an object, use `ALTER`:

```sql
ALTER MATERIALIZED VIEW winning_bids SET (RETAIN HISTORY FOR '2hr');
```

> <!-- TODO(maddyblue): replace this paragraph with a mention of the catalog

Reviewer: This exists now https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_history_retention_strategies

Author: The PR just barely missed the release cutoff b/c it was cut early for the onsite :( I'll put a warn if unreleased.

Reviewer: aww, rip

     table/column once it's available -->
To see what history retention period has been configured for an object, run the
`SHOW CREATE ...` statement for the given object. See [`SHOW CREATE SOURCE`](/sql/show-create-source/)
for an example.

```sql
SHOW CREATE MATERIALIZED VIEW winning_bids;
```
```nofmt
              name               |                                                                                                                       create_sql
---------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 materialize.public.winning_bids | CREATE MATERIALIZED VIEW "materialize"."public"."winning_bids" IN CLUSTER "quickstart" WITH (RETAIN HISTORY = FOR '2hr') AS SELECT * FROM "materialize"."public"."highest_bid_per_auction" WHERE "end_time" < "mz_catalog"."mz_now"()
```

### Disabling history retention

> To disable history retention, set the `RETAIN HISTORY` option to its original
> default value (1 second):

Reviewer: ```suggestion
To disable history retention, reset the `RETAIN HISTORY` option:
```

Author: Agree to your suggestion - one wrinkle is that `mz_internal.mz_history_retention_strategies` returns 
```
strategy |  value  
---------+---------
FOR      | 1000
```
when retain history is unset. Is it confusing to say everywhere "no retain history", but then this table doesn't say zero it says 1 second? Should we put a small note somewhere that the default/"no retain history" == 1 second

Reviewer: I had similar concerns when writing this, and I think that we do want to expose exactly what the current retain history is in ms. We could document this somewhere. We also could add a third column `is_default` or something that's a bool? I'm fine with any solution here.

```sql

> ALTER MATERIALIZED VIEW winning_bids SET (RETAIN HISTORY FOR '1s');

Reviewer: ```suggestion
ALTER MATERIALIZED VIEW winning_bids RESET (RETAIN HISTORY);
```

Author: Ah I missed this syntax. Need to add it to RR diagrams etc

### Resource utilization

It's important to note that increasing the history retention period for an
object will lead to increased resource utilization in Materialize.

* **For sources, tables and materialized views:** increasing the history
    retention period for these objects increases the amount of historical data
    that is retained in the storage layer. You can expect storage resource
    utilization to increase, which may incur additional costs.

* **For indexes:** increasing the history retention period for an index
    increases the amount of historical data that is retained in memory in the
    cluster maintaining the index. You can expect memory resource utilization
    for that cluster to go up, which might require you to size up your cluster
    and consume additional compute credits.

### Clean-up

The history retention period represents the minimum amount of historical data
guaranteed to be retained by Materialize. History clean-up is processed in the
background, so older history may be accessible for the period of time between
when it falls outside the retention period and when it is cleaned up.

### Best practices

> Given the additional memory costs associated with increasing the history
> retention period on indexes, if you don't need the performance benefits of an
> index—in particular for the use case of durable subscriptions-you should consider creating a materialized view for your subscription query
> and configure the history retention period on that materialized view
> instead.

Reviewer 2: ```suggestion
Given the additional memory costs associated with increasing the history
retention period on indexes, if you don't need the performance benefits of an
index—in particular for the use case of durable subscriptions-you should
consider creating a materialized view for your subscription query and configure
the history retention period on that materialized view instead.
```

In the case that your application crashes and you need to resume your subscription using the
persisted progress message `mz_timestamp`:
* If you were processing data in your application before persisting the subsequent progress message's `mz_timestamp`: you may end up processing duplicate data.
* If you were persisting the progress message's `mz_timestamp` before processing all the
buffered data from before that progress message: you may end up dropping some data.

As a result, the way to guarantee exactly once data processing in the case of your application
crashes is to write the progress message `mz_timestamp` and all buffered data together in a single transaction.

<!--TODO(chaas): add top-level section on point-in-time queries, similar to Durable
subscriptions section-->

## Examples

### Enabling history retention

To set a history retention period for an object, use the `RETAIN HISTORY`
option, which accepts positive [interval](/sql/types/interval/) values
(e.g. `'1hr'`):

```sql
CREATE MATERIALIZED VIEW winning_bids
WITH (RETAIN HISTORY FOR '1hr') AS
SELECT auction_id,
       bid_id,
       item,
       amount
FROM highest_bid_per_auction
WHERE end_time < mz_now();
```

> ### Adjusting history retention configuration
> To adjust the history retention period for an object, use `ALTER`:

Reviewer 2: ```suggestion
### Adjusting history retention configuration

To adjust the history retention period for an object, use `ALTER`:
```

Reviewer 2: Sorry for being a sociopath.

```sql
ALTER MATERIALIZED VIEW winning_bids SET (RETAIN HISTORY FOR '2hr');
```

> ### Viewing history retention configuration
> To see what history retention period has been configured for an object,

Reviewer 2: ```suggestion
### Viewing history retention configuration

To see what history retention period has been configured for an object,
```

1. As results come in continuously, buffer the latest results in memory until you receive
a [progress](/sql/subscribe#progress) message. At that point, the data up until the progress message
is complete, so you can:

   1. Process all the buffered data in your application.
   1. Persist the `mz_timestamp` of the progress message.

1. To resume the subscription in subsequent restarts,
use the following continuous query against Materialize in your application code:

   ```sql
   SUBSCRIBE (<your query>) WITH (PROGRESS, SNAPSHOT false) AS OF <last_progress_mz_timestamp>;
   ```

   In a similar way, as results come in continuously, buffer the latest results
   in memory until you receive a [progress](/sql/subscribe#progress) message. At that point,
   the data up until the progress message is complete, so you can:

   1. Process all the buffered data in your application.
   1. Persist the `mz_timestamp` of the progress message.

   You can tweak the flush interval at which you durably record the latest
   progress timestamp, if every progress message is too frequent.

<!--TODO(chaas): add section on DECLARE + FETCH as a way to get around continuous connection -->

> ### Idempotency

Reviewer: This new section LGTM. I like that we give a simple-enough framework, and provide guidance on what do to for a better solution.
